app/models/grouping.rb

```diff
@@ -728,8 +728,28 @@ def get_next_grouping(current_role, reversed)
     next_grouping
   end
 
+  def get_next_random_unmarked(current_role)
+    groupings = get_all_unmarked(current_role)
+    groupings.order('RANDOM()').first
+  end
+
   private
 
+  def get_all_unmarked(current_role)
+    if current_role.ta?
+      # Get relevant groupings for a TA
+      groupings = current_role.groupings.joins(:current_result)
```

Comment:
This is quite subtle, but `joins(:current_result)` doesn't actually work. Instead you should use `Assignment#current_results` and go from there.